diff --git a/.classpath b/.classpath
index 56e6d3b..642ca32 100644
--- a/.classpath
+++ b/.classpath
@@ -1,42 +1,60 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
-	<classpathentry kind="src" output="target/classes" path="src/main/java">
+	<classpathentry excluding="org/hyperledger/fabric/protos/**/*.java" kind="src" output="target/classes" path="src/main/java">
 		<attributes>
 			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="src" output="target/classes" path="target/generated-sources/protobuf/java">
+	<classpathentry kind="src" output="target/test-classes" path="src/test/java">
 		<attributes>
 			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
+			<attribute name="test" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="src" output="target/classes" path="target/generated-sources/protobuf/grpc-java">
+	<classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
 		<attributes>
-			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="src" output="target/test-classes" path="src/test/java">
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry excluding="**" kind="src" output="target/classes" path="src">
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="src" path="target/generated-sources/annotations">
 		<attributes>
 			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
+			<attribute name="ignore_optional_problems" value="true"/>
+			<attribute name="m2e-apt" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
+	<classpathentry excluding="org/hyperledger/fabric/protos/**/*.java" kind="src" output="target/classes" path="target/generated-sources/protobuf/java">
 		<attributes>
+			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8">
+	<classpathentry excluding="org/hyperledger/fabric/protos/**/*.java" kind="src" output="target/classes" path="target/generated-sources/protobuf/grpc-java">
 		<attributes>
+			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry excluding="**" kind="src" output="target/classes" path="src">
+	<classpathentry kind="src" output="target/test-classes" path="target/generated-test-sources/test-annotations">
 		<attributes>
+			<attribute name="optional" value="true"/>
 			<attribute name="maven.pomderived" value="true"/>
+			<attribute name="ignore_optional_problems" value="true"/>
+			<attribute name="m2e-apt" value="true"/>
+			<attribute name="test" value="true"/>
 		</attributes>
 	</classpathentry>
 	<classpathentry kind="output" path="target/classes"/>
diff --git a/.project b/.project
index d4e2cd7..e154f30 100644
--- a/.project
+++ b/.project
@@ -11,12 +11,12 @@
 			</arguments>
 		</buildCommand>
 		<buildCommand>
-			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<name>net.sf.eclipsecs.core.CheckstyleBuilder</name>
 			<arguments>
 			</arguments>
 		</buildCommand>
 		<buildCommand>
-			<name>net.sf.eclipsecs.core.CheckstyleBuilder</name>
+			<name>org.eclipse.m2e.core.maven2Builder</name>
 			<arguments>
 			</arguments>
 		</buildCommand>
diff --git a/pom.xml b/pom.xml
index 9042e8c..d144c1f 100644
--- a/pom.xml
+++ b/pom.xml
@@ -99,7 +99,7 @@
         <dependency>
             <groupId>io.netty</groupId>
             <artifactId>netty-tcnative-boringssl-static</artifactId>
-            <version>2.0.34.Final</version>
+            <version>2.0.35.Final</version>
         </dependency>
         <dependency>
             <groupId>io.netty</groupId>
@@ -642,13 +642,13 @@
                             </execution>
                         </executions>
                     </plugin>
-                    <plugin>
+                    <!---<plugin>
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-gpg-plugin</artifactId>
                         <configuration>
                             <useAgent>true</useAgent>
                         </configuration>
-                        <version>1.6</version>  <!-- was 1.5 -->
+                        <version>1.6</version>  <was 1.5>
                         <executions>
                             <execution>
                                 <id>sign-artifacts</id>
@@ -658,7 +658,7 @@
                                 </goals>
                             </execution>
                         </executions>
-                    </plugin>
+                    </plugin>-->
                     <plugin>
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-source-plugin</artifactId>
diff --git a/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixCredential.java b/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixCredential.java
index 7d57964..87f4343 100644
--- a/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixCredential.java
+++ b/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixCredential.java
@@ -161,7 +161,7 @@ public class IdemixCredential {
     /**
      * @return A proto representation of this credential
      */
-    Idemix.Credential toProto() {
+    public Idemix.Credential toProto() {
         Idemix.Credential.Builder builder = Idemix.Credential.newBuilder()
                 .setA(IdemixUtils.transformToProto(A))
                 .setB(IdemixUtils.transformToProto(B))
@@ -174,4 +174,4 @@ public class IdemixCredential {
 
         return builder.build();
     }
-}
\ No newline at end of file
+}
diff --git a/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixIssuerPublicKey.java b/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixIssuerPublicKey.java
index a5e3359..408f362 100644
--- a/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixIssuerPublicKey.java
+++ b/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixIssuerPublicKey.java
@@ -197,7 +197,7 @@ public class IdemixIssuerPublicKey {
     /**
      * @return A proto version of this issuer public key
      */
-     Idemix.IssuerPublicKey toProto() {
+     public Idemix.IssuerPublicKey toProto() {
 
         Idemix.ECP[] ipkHAttrs = new Idemix.ECP[HAttrs.length];
         for (int i = 0; i < HAttrs.length; i++) {
diff --git a/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixPseudonym.java b/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixPseudonym.java
index f38210d..6c8459c 100644
--- a/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixPseudonym.java
+++ b/src/main/java/org/hyperledger/fabric/sdk/idemix/IdemixPseudonym.java
@@ -54,7 +54,7 @@ public class IdemixPseudonym {
     /**
      * @return the secret randomness used to construct this pseudonym
      */
-     BIG getRandNym() {
+     public BIG getRandNym() {
         return RandNym;
     }
 }
diff --git a/src/main/java/org/hyperledger/fabric/sdk/transaction/TransactionContext.java b/src/main/java/org/hyperledger/fabric/sdk/transaction/TransactionContext.java
index 012238d..9a0a92c 100644
--- a/src/main/java/org/hyperledger/fabric/sdk/transaction/TransactionContext.java
+++ b/src/main/java/org/hyperledger/fabric/sdk/transaction/TransactionContext.java
@@ -246,4 +246,8 @@ public class TransactionContext {
     public Identities.SerializedIdentity getSerializedIdentity() {
         return identity;
     }
+
+    public SigningIdentity getSigningIdentity() {
+        return signingIdentity;
+    }
 }  // end TransactionContext
